pkg/mei/mei.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mei
+
+import (
+	"encoding/binary"
+	"fmt"
+	"os"
+	"syscall"
+	"unsafe"
+)
+
+// DefaultMEIDevicePath is the path of the default MEI device. This file will be
+// present if the "mei_me" kernel module is loaded.
+var DefaultMEIDevicePath = "/dev/mei0"
+
+// HECIGuids maps the known HECI GUIDs to their values. The MEI interface wants
+// little-endian. See all the GUIDs at
+// https://github.com/intel/lms/blob/master/MEIClient/Include/HECI_if.h
+var (
+	// "8e6a6715-9abc-4043-88ef-9e39c6f63e0f"
+	MKHIGuid = ClientGUID{0x15, 0x67, 0x6a, 0x8e, 0xbc, 0x9a, 0x43, 0x40, 0x88, 0xef, 0x9e, 0x39, 0xc6, 0xf6, 0x3e, 0xf}
+)
+
+// see include/uapi/linux/mei.h
+var (
+	IoctlMEIConnectClient = IOWR('H', 0x01, uintptr(len(ClientGUID{})))
+)
+
+// ClientGUID is the data buffer to pass to `ioctl` to connect to
+// MEI. See include/uapi/linux/mei.h .
+type ClientGUID [16]byte
+
+// ClientProperties is the data buffer returned by `ioctl` after connecting to
+// MEI. See include/uapi/linux/mei.h .
+type ClientProperties [6]byte
+
+// MaxMsgLength is the maximum size of a message for this client.
+func (c ClientProperties) MaxMsgLength() uint32 {
+	return binary.LittleEndian.Uint32(c[:4])
+}
+
+// ProtocolVersion is this client's protocol version.
+func (c ClientProperties) ProtocolVersion() uint8 {
+	return c[4]
+}
+
+// MEI represents an Intel ME Interface object.
+type MEI struct {
+	fd *os.File
+}
+
+// Open opens the specified MEI device, using the client type defined by the
+// given name. See `HECIGuids` in this package.
+func (m *MEI) Open(meiPath string, guid ClientGUID) (*ClientProperties, error) {
```

Comment:
I'm not sure I fully understand the MEI API here. Could you do something like OpenMEI(...) (*MEI, *ClientProperties, error)? Or fold ClientProperties into MEI?

then you don't have the race with fd either.

Reply:
yeah, reading further it seems to me like you could fold it into MEI, esp since it's associated specifically with that fd?

Reply:
the client properties are client-specific (e.g. the MKHI client, basically the MEI interface lets you use different client types based on what GUID is passed via ioctl when opening the fd). I could fold it into MEI, but being MKHI-specific I'd rather leave it here. Alternatively I can move it to MEI, and access it via `m.MEI.clientProperties.SomeMethod()` instead of `m.clientProperties.SomeMethod()`

Reply:
But you derive in MEI.Open, not in mkhi.open? That led me to believe they're MEI specific.

Reply:
I double-checked the implementation, and they are MEI-specific, not MKHI-specific, sorry about the confusion. Yeah, I'll fold them into `MEI`

Reply:
done in the latest push